coverage/PENDING_SUBMIT/intel/0503-COVERAGE-anv-nir-apply-pipeline-layout-560/cov-loop-condition-clamp-vec-of-ones.amber

```diff
@@ -0,0 +1,283 @@
+#!amber
+
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+# A test for a coverage-gap found by the GraphicsFuzz project.
+
+# Short description: A fragment shader that covers specific NIR code paths
+
+# The test passes because the shader always writes red.
+
+# Optimized using spirv-opt with the following arguments:
+# '--scalar-replacement=100'
+# '--eliminate-dead-inserts'
+# '--eliminate-local-multi-store'
+# '--convert-local-access-chains'
+# '--eliminate-dead-branches'
+# '--merge-blocks'
+# '--eliminate-dead-branches'
+# '--eliminate-dead-branches'
+# '--merge-blocks'
+# '--scalar-replacement=100'
+# '--eliminate-dead-branches'
+# '--merge-blocks'
+# '--ccp'
+# '--eliminate-local-single-store'
+# '--scalar-replacement=100'
+# '--convert-local-access-chains'
+# '--ccp'
+# '--reduce-load-size'
+# '--eliminate-dead-branches'
+# '--eliminate-local-single-block'
+# '--redundancy-elimination'
+# '--combine-access-chains'
+# '--ccp'
+# '--inline-entry-points-exhaustive'
+# '--scalar-replacement=100'
+# '--ccp'
+# spirv-opt commit hash: a0370efd589be33d5d9a85cfde2f85841b3755af
+
+
+
+SHADER vertex variant_vertex_shader PASSTHROUGH
+
+# variant_fragment_shader is derived from the following GLSL:
+# #version 320 es
+# #define _int_1 _GLF_uniform_int_values[0]
+# #define _int_0 _GLF_uniform_int_values[1]
+# #define _float_1_0 _GLF_uniform_float_values[0]
+# #define _float_0_5 _GLF_uniform_float_values[1]
+#
+# precision highp float;
+# precision highp int;
+#
+# // Contents of _GLF_uniform_int_values: [1, 0]
+# layout(set = 0, binding = 0) uniform buf0
+# {
+#     int _GLF_uniform_int_values[2];
+# };
+#
+# // Contents of _GLF_uniform_float_values: [1.0, 0.5]
+# layout(set = 0, binding = 1) uniform buf1
+# {
+#     float _GLF_uniform_float_values[2];
+# };
+#
+# // Contents of one: [1.0, 1.0]
+# layout(set = 0, binding = 2) uniform buf2
+# {
+#     vec2 one;
+# };
+#
+# layout(location = 0) out vec4 _GLF_color;
+#
+# void main()
+# {
+#     float f = _float_1_0;
+#
+#     // Iterated once.
+#     do
+#     {
+#         // Iterated once.
+#         for (int i = 0; i != clamp(_int_0, int(one.x), 2); i++)
+#         {
+#             // Always chooses 2.0, thus f becomes 0.5.
+#             f /= one.y > _float_1_0 ? 256.0 : 2.0;
```

Comment:
Vulkan SPIR-V allows division to have some roundoff error:

https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#spirvenv-precision-operation

OpFDiv(x,y) is allowed 2.5 ULPs when |y| is in the range [2-126, 2126].

Thus f / one.y *could* end up being slightly larger than 1.0 if this round-off is used.

How about changing this to:

`f /= one.y > _float_1.0 + 0.1 ? 256.0 : 2.0`

?

Would that still get the coverage?

We did once have a SwiftShader "bug" that turned out not to be a bug due to a problem with dividing by 1.0., so I'm a little nervous.

Reply:
I think you got the precedence wrong. The right hand side is calculated first so you can think of it as `f /= (one.y > _float_1_0 ? 256.0 : 2.0)`. So all the line does is f /= 2.0. We could add some threshold to the result checking though.